### ISSUE TITLE
Three parameters date functions have unpredictable behavior

### ISSUE THREAD

**Issue by astefan** [tableau/connector-plugin-sdk#305]
I am testing a JDBC connector I'm working on and the three-parameters date functions doesn't seem to be taken into consideration somehow and the TDVT tests fall back (or this is what I'm assuming based on the tests results) to the two parameters versions.

This was tested on the latest SDK checked out few hours ago and Tableau 2019.3.

My `dialect.tdd` file header and the overridden function is the following (and has many other functions definitions).

```
<?xml version="1.0" encoding="utf-8"?>
<dialect name='Custom'
         class='custom_jdbc'
         version='18.1'>
         base='SQL92Dialect'>
   <function-map>
....
<date-function name='DATEPART' return-type='int'>
      <formula>DATE_PART(%1,%2)</formula>
      <formula part='year'>DATE_PART(&apos;year&apos;,%2)</formula>
      <formula part='quarter'>DATE_PART(&apos;quarter&apos;,%2)</formula>
      <formula part='month'>DATE_PART(&apos;month&apos;,%2)</formula>
      <formula part='dayofyear'>DATE_PART(&apos;dayofyear&apos;,%2)</formula>
      <formula part='day'>DATE_PART(&apos;day&apos;,%2)</formula>
      <formula part='weekday'>DATE_PART(&apos;weekday&apos;,%2)</formula>
      <formula part='week'>DATE_PART(&apos;week&apos;,%2)</formula>
      <formula part='hour'>DATE_PART(&apos;hour&apos;,%2)</formula>
      <formula part='minute'>DATE_PART(&apos;minute&apos;,%2)</formula>
      <formula part='second'>DATE_PART(&apos;second&apos;,%2)</formula>
      <argument type='localstr' />
      <argument type='datetime' />
    </date-function>
    <date-function name='DATEPART' return-type='int'>
      <formula>DATE_PART(%1,%2)</formula>
      <formula part='year'>DATE_PART(&apos;year&apos;,%2)</formula>
      <formula part='quarter'>DATE_PART(&apos;quarter&apos;,%2)</formula>
      <formula part='month'>DATE_PART(&apos;month&apos;,%2)</formula>
      <formula part='dayofyear'>DATE_PART(&apos;dayofyear&apos;,%2)</formula>
      <formula part='day'>DATE_PART(&apos;day&apos;,%2)</formula>
      <formula part='weekday'>DATE_PART(&apos;weekday&apos;,%2)</formula>
      <formula part='week'>DATE_PART(&apos;week&apos;,%2)</formula>
      <formula part='hour'>DATE_PART(&apos;hour&apos;,%2)</formula>
      <formula part='minute'>DATE_PART(&apos;minute&apos;,%2)</formula>
      <formula part='second'>DATE_PART(&apos;second&apos;,%2)</formula>
      <argument type='localstr' />
      <argument type='date' />
    </date-function>
    <date-function name='DATEPART' return-type='int'>
      <formula part='year'>DATE_PART(&apos;year&apos;,%2)</formula>
      <formula part='quarter'>DATE_PART(&apos;quarter&apos;,%2)</formula>
      <formula part='month'>DATE_PART(&apos;month&apos;,%2)</formula>
      <formula part='dayofyear'>DATE_PART(&apos;dayofyear&apos;,%2)</formula>
      <formula part='day'>DATE_PART(&apos;day&apos;,%2)</formula>
      <formula part='weekday'>IIF(%3 = 'Monday', ISO_DAY_OF_WEEK(%2), DATE_PART(&apos;weekday&apos;,%2))</formula>
      <formula part='week'>IIF(%3 = 'Monday', ISO_WEEK_OF_YEAR(%2), DATE_PART(&apos;weekday&apos;,%2))</formula>
      <formula part='hour'>DATE_PART(&apos;hour&apos;,%2)</formula>
      <formula part='minute'>DATE_PART(&apos;minute&apos;,%2)</formula>
      <formula part='second'>DATE_PART(&apos;second&apos;,%2)</formula>
      <argument type='localstr' />
      <argument type='date' />
      <argument type='localstr' />
    </date-function>
    <date-function name='DATEPART' return-type='int'>
      <formula part='year'>DATE_PART(&apos;year&apos;,%2)</formula>
      <formula part='quarter'>DATE_PART(&apos;quarter&apos;,%2)</formula>
      <formula part='month'>DATE_PART(&apos;month&apos;,%2)</formula>
      <formula part='dayofyear'>DATE_PART(&apos;dayofyear&apos;,%2)</formula>
      <formula part='day'>DATE_PART(&apos;day&apos;,%2)</formula>
      <formula part='weekday'>IIF(%3 = 'Monday', ISO_DAY_OF_WEEK(%2), DATE_PART(&apos;weekday&apos;,%2))</formula>
      <formula part='week'>IIF(%3 = 'Monday', ISO_WEEK_OF_YEAR(%2), DATE_PART(&apos;weekday&apos;,%2))</formula>
      <formula part='hour'>DATE_PART(&apos;hour&apos;,%2)</formula>
      <formula part='minute'>DATE_PART(&apos;minute&apos;,%2)</formula>
      <formula part='second'>DATE_PART(&apos;second&apos;,%2)</formula>
      <argument type='localstr' />
      <argument type='datetime' />
      <argument type='localstr' />
    </date-function>
....
</function-map>
</dialect>
```

As you can see in the functions definitions, the three-parameters ones have special definitions fro `week` and `weekday` where the formula chooses between one function or another based on the value of the third parameter.

Looking in the tests' results, for `exprtests/standard\setup.date.datepart.sow.week.txt` which should test `DATEPART('week', [date2], 'monday')`, the actual query being created is
```
SELECT DATE_PART(,"calcs"."date2") AS "TEMP_Test________________"
FROM "calcs"
GROUP BY 1
```

**My `IIF` is not there and, even more, the first parameter is not passed at all.** I can only guess, that it falls back to the generic formula for the two-parameters version of the function (`DATE_PART(%1,%2)`).

---

**In a second tes**t, if I define the generic formulas for the two-parameters functions definitions to be as `<formula>DATE_PART(&apos;%1&apos;,%2)</formula>` for the same test - `DATEPART('week', [date2], 'monday')` - the query being generated is:

```
SELECT DATE_PART('',"calcs"."date2") AS "TEMP_Test________________"
FROM "calcs"
GROUP BY 1
```

The only difference is the single quotes as the first parameter. Strange is that they are empty and `week` is not passed as a parameter. But still, my special formula having `IIF` in it is not used at all. It's like the three-parameters functions are ignored and Tableau falls back to the two-parameters functions.

---

**In a third test**, I also defined generic formulas - `<formula>DATE_PART(%1,%2,%3)</formula>` - for the three-parameters functions as well with a noticeable change in the tests results. The query created for was

```
SELECT DATE_PART(,"calcs"."date2",1) AS "TEMP_Test________________"
FROM "calcs"
GROUP BY 1
```
which seems to indicate that `monday` is actually value `1` (and not `monday`) and, again, the first parameter - the "part" - doesn't seem to have a value.

**What's also confusing to me is that the same test, but with `sunday` value as third parameter** - `DATEPART('week', [date2], 'sunday')` - does create a valid query: 

```
SELECT DATE_PART('week',"calcs"."date2") AS "TEMP_Test________________"
FROM "calcs"
GROUP BY 1
```

Again, there's no `IIF` in my query and for this one I'm assuming it doesn't use the generic formula, but one that I defined explicitly for the two-parameters functions and "part" `week`.

---

**An interesting test was with a generic formula-only defined for the three params functions**:

```
    <date-function name='DATEPART' return-type='int'>
      <formula>CASE WHEN &apos;%3&apos; = '1' AND &apos;%1&apos; = 'week' THEN ISO_DAY_OF_WEEK(%2) WHEN &apos;%3&apos; = '1' AND &apos;%1&apos; = 'weekday' THEN ISO_WEEK_OF_YEAR(%2) ELSE DATE_PART(&apos;%1&apos;,%2) END</formula>
      <argument type='localstr' />
      <argument type='date' />
      <argument type='localstr' />
    </date-function>
```

In this case, my **custom formula was used, but the "part" parameter wasn't passed along**. This is the query that was created for `DATEPART('week', [date2], 'monday')` test

```
      SELECT CASE WHEN '1' = '1' AND '' = 'week' THEN ISO_DAY_OF_WEEK("calcs"."date2") WHEN '1' = '1' AND '' = 'weekday' THEN ISO_WEEK_OF_YEAR("calcs"."date2") ELSE DATE_PART('',"calcs"."date2") END AS "TEMP_Test________________"
FROM "calcs"
GROUP BY 1
```

As mentioned before, "monday" has the value `1` and the "part" is missing completely. And I am stuck here and cannot explain all this behavior.

**Comment by lriggs** [comment]
Thank you for your detailed notes. I was able to reproduce the problem and will take a look into it.

**Comment by astefan** [comment]
If it helps, now I'm playing with DATEDIFF and I'm seeing the same behavior. I haven't tested all of datetime functions, but at least for three of them the issue is there. Might be all, not sure, just wanted to mention this.

**Comment by lriggs** [comment]
I found a few things in my testing. First, there is a bug in 2019.3 that means Tableau isn't always picking the right function (as you noted). This is fixed in our upcoming release for 2019.4 which should be available soon.

I also learned that Sunday is treated as the start of the week, so any of the three part functions that use Sunday will default back to the two part version.

Also, only the 'week' part for the three argument functions is actually relevant. The other parts aren't used since the only Start of Week function we support is for 'week'. Tableau will pick the 2 argument function for other date parts.

I think in 2019.4 things will work as expected. You might be able to get a beta account in order to test that, or it should be available sometime soon, although I don't know exactly when.
